### PR TITLE
Upgrade libsodium for AVX2-detection bugfix.

### PR DESCRIPTION
**Pull request by defuse**
Upgrades libsodium to bring in an AVX2-detection bugfix (to help with #1130).

### CODE REVIEW

**Comment by daira** [comment]
utACK.

**Comment by ebfull** [comment]
ACK

@zkbot r+

**Comment by defuse** [comment]
Interesting failures... I'll have a look.

**Comment by defuse** [comment]
libsodium does the s < L check in the new version, so @daira and I removed our own check and added a runtime check that libsodium is actually checking that (in case we somehow get linked against an older version of libsodium).

**Comment by defuse** [comment]
We should test this at least once with ED25519_COMPAT defined to make sure our runtime test works. https://github.com/jedisct1/libsodium/commit/62911edb7ff2275cccd74bf1c8aefcc4d76924e0

**Comment by defuse** [comment]
I'm trying to ad -DED25519_COMPAT and it's not working :(

**Comment by defuse** [comment]
Got it to work:


```
[firexware@firexware-pc zcash]$ git diff
diff --git a/depends/packages/libsodium.mk b/depends/packages/libsodium.mk
index d7717bb..70728c3 100644
--- a/depends/packages/libsodium.mk
+++ b/depends/packages/libsodium.mk
@@ -15,7 +15,7 @@ define $(package)_config_cmds
 endef
 
 define $(package)_build_cmds
-  $(MAKE)
+   $(MAKE) CPPFLAGS="-DED25519_COMPAT"
 endef
 
 define $(package)_stage_cmds
```

```
[firexware@firexware-pc zcash]$ ./src/zcash-gtest 
zcash-gtest: ./crypto/common.h:105: int init_and_check_sodium(): Assertion `crypto_sign_verify_detached(sig, message, sizeof(message), pk) != 0' failed.
Aborted (core dumped)
```

**Comment by daira** [comment]
ACK, coverage not checked.

**Comment by daira** [comment]
@zkbot try

**Comment by ebfull** [comment]
Wow, `try` actually worked?

File: src/crypto/common.h [hunk 1]
```diff
@@ -63,4 +69,42 @@ void static inline WriteBE64(unsigned char* ptr, uint64_t x)
...
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10 };
+
+    // Add L to S, which starts at sig[32].
+    unsigned int s = 0;
+    for (size_t i = 0; i < 32; i++) {
+        s = sig[32 + i] + L[i] + (s >> 8);
+        sig[32 + i] = s & 0xff;
+    }
+
+    assert(crypto_sign_verify_detached(sig, message, sizeof(message), pk) != 0);
```
Comment (defuse) [inline src/crypto/common.h]: The "== 0" case won't be covered, but I manually tested it as you can see in the comments on this PR.

Comment (defuse) [inline src/crypto/common.h]: It's infeasible to add a unit test that tests that branch, as it requires passing options to libsodium's build. By the time we've built the unit tests, it's too late to do that.

**Comment by defuse** [comment]
I think the test coverage that's lacking is acceptable, since to follow some branches you need to rebuild libsodium with different options. I tested that manually in https://github.com/zcash/zcash/pull/1268#issuecomment-240881304

**Comment by defuse** [comment]
Can we r+ this now?

**Comment by ebfull** [comment]
ACK

@zkbot r+

**Comment by ebfull** [comment]
There we go, `try` at work!

@zkbot retry

**Pull request merged**